### ISSUE TITLE
Alpha 23: _zipTypesAndAnnotaions throws exception -> no DI possible ! 

### ISSUE THREAD

**Issue by jomeier** [angular/angular#1944]
Hi,

I get exceptions in the dependency injection code of Angular alpha 23.

In angular2/src/reflection/reflection_capabilities.ts I always get an error message if I use dependency injection with typscript 1.5beta. In a debugging session I found this code snippet:

```javascript
if (isPresent(paramTypes) || isPresent(paramAnnotations)) {
    return this._zipTypesAndAnnotaions(paramTypes, paramAnnotations);
}
```

In my case paramTypes is undefined but paramAnnotations contains an array and cause of the || _zipTypesAndAnnotaions (a typo? should be named ...AndAnnotations I think) will be called.

In _zipTypesAndAnnotaions this line causes an exception in my demo application:

```javascript
_zipTypesAndAnnotaions(paramTypes, paramAnnotations): List<List<any>> {
    var result = ListWrapper.createFixedSize(paramTypes.length);
```

cause paramTypes is undefined.

In my opinion it should check if paramTypes is undefined and then take paramAnnotations for creating the result list.

Here's an excerpt of my code which shows how I tried to use dependency injection:

```javascript
class test1 {
    gets() { return "balupp";}
}


// Annotation section
@Component({
    selector: 'app',
    injectables: [test1]
})
@View({
...
})

// Component controller
class MyAppComponent {
    constructor(@Inject(test1) webs) {
        ...
    }
```

Best regards,

Josef

**Comment by jomeier** [comment]
This is the function with a few of my modifications. Mostly I check if paramTypes is undefined. If so, paramAnnotations will be used for the length of the result array.
If paramTypes is undefined, the result array will be filled with the values from paramAnnotations:

```javascript
_zipTypesAndAnnotaions: function (paramTypes, paramAnnotations) {
            var result;
            if( typeof paramTypes === 'undefined' )
               result = ListWrapper.createFixedSize(paramAnnotations.length);
            else
               result = ListWrapper.createFixedSize(paramTypes.length);

            for (var i = 0; i < result.length; i++) {
               if (typeof paramTypes === "undefined")
                 result[i] = [];
               else
                 if (paramTypes[i] != Object) {
                result[i] = [paramTypes[i]];
              } else {
                result[i] = [];
              }
              if (isPresent(paramAnnotations) && isPresent(paramAnnotations[i])) {
                result[i] = result[i].concat(paramAnnotations[i]);
              }
            }
            return result;
          },
```

Could someone acknowledge that this is the solution to a bug, please?

**Comment by vicb** [comment]
please close if fixed by #1955

**Comment by jomeier** [comment]
```javascript 
<PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
    <DebugSymbols>true</DebugSymbols>
    <TypeScriptRemoveComments>false</TypeScriptRemoveComments>
    <TypeScriptSourceMap>true</TypeScriptSourceMap>
    <TypeScriptAdditionalFlags> $(TypeScriptAdditionalFlags) --emitDecoratorMetadata </TypeScriptAdditionalFlags>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
    <DebugSymbols>true</DebugSymbols>
    <TypeScriptRemoveComments>true</TypeScriptRemoveComments>
    <TypeScriptSourceMap>false</TypeScriptSourceMap>
    <TypeScriptAdditionalFlags> $(TypeScriptAdditionalFlags) --emitDecoratorMetadata </TypeScriptAdditionalFlags>
  </PropertyGroup>
```
After putting this in my njsproj file dependency injection works like a charme :-) Thanks a lot!

Josef
